File: backend/routers/terms.py
"""
词条（Terms）路由
=================
处理所有 /terms 相关的 HTTP 请求。
只负责 HTTP 关注点：解析参数、调用 crud、返回响应或 404。
不写 SQLAlchemy 查询，不做 tags 转换。
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from backend import crud
from backend.database import get_db
from backend.schemas import TermCreate, TermResponse, TermStatus, TermUpdate

router = APIRouter(prefix="/terms", tags=["terms"])


@router.post("", response_model=TermResponse, status_code=201)
def create_term(data: TermCreate, db: Session = Depends(get_db)):
    return crud.create_term(db, data)


@router.get("", response_model=list[TermResponse])
def list_terms(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    status: TermStatus | None = None,
    tag: str | None = None,
    db: Session = Depends(get_db),
):
    return crud.list_terms(db, skip=skip, limit=limit, status=status, tag=tag)


@router.get("/{term_id}", response_model=TermResponse)
def get_term(term_id: int, db: Session = Depends(get_db)):
    term = crud.get_term(db, term_id)
    if term is None:
        raise HTTPException(status_code=404, detail="Term not found")
    return term


@router.put("/{term_id}", response_model=TermResponse)
def update_term(term_id: int, data: TermUpdate, db: Session = Depends(get_db)):
    term = crud.update_term(db, term_id, data)
    if term is None:
        raise HTTPException(status_code=404, detail="Term not found")
    return term


@router.delete("/{term_id}")
def delete_term(term_id: int, db: Session = Depends(get_db)):
    deleted = crud.delete_term(db, term_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Term not found")
    return {"detail": "Term deleted"}

File: frontend/src/api/terms.ts
/**
 * Terms API
 * =========
 * 对应后端路由 /terms (routers/terms.py)。
 *
 * 规则：只有 src/api/ 下的文件才能 import axios/apiClient。
 * Pages 和 components 通过 hooks/ 里的 TanStack Query 钩子访问数据。
 */

import type { Term, TermCreate, TermUpdate, TermsQueryParams } from '@/types/api'
import apiClient from './client'

/**
 * GET /terms
 * 获取词条列表，支持分页、状态过滤和标签过滤。
 */
export async function getTerms(params?: TermsQueryParams): Promise<Term[]> {
  const response = await apiClient.get<Term[]>('/terms', { params })
  return response.data
}

/**
 * GET /terms/:id
 * 获取单个词条，不存在时后端返回 404。
 */
export async function getTerm(id: number): Promise<Term> {
  const response = await apiClient.get<Term>(`/terms/${id}`)
  return response.data
}

/**
 * POST /terms
 * 创建词条。只有 term 字段是必填的。
 */
export async function createTerm(data: TermCreate): Promise<Term> {
  const response = await apiClient.post<Term>('/terms', data)
  return response.data
}

/**
 * PUT /terms/:id
 * 局部更新词条。未传的字段保持不变；tags: [] 会清空标签。
 */
export async function updateTerm(id: number, data: TermUpdate): Promise<Term> {
  const response = await apiClient.put<Term>(`/terms/${id}`, data)
  return response.data
}

/**
 * DELETE /terms/:id
 * 删除词条及其所有复习记录（级联删除）。
 * 后端返回 { "detail": "Term deleted" }。
 */
export async function deleteTerm(id: number): Promise<{ detail: string }> {
  const response = await apiClient.delete<{ detail: string }>(`/terms/${id}`)
  return response.data
}
